design-proposals/mass_machine_type_transition.md

```diff
@@ -0,0 +1,265 @@
+option to allow the job to automatically restart every running VM that is being
+updated and apply the changes immediately.
+
+If the Job fails or is killed, it will be restarted with `restartPolicy=OnFailure`
+in the Job spec. The `restart-vm-required` label may then remain on a VM that
+has already been shut down, so the Job will first check VMs with the label and
+confirm if the machine types have been updated. If they have, then the label
+will be removed, but if not, the label will remain and the VM will continue to
+be tracked until it is restarted and its machine type is updated.
+
+The user may use the `--which-matches-glob` flag to specify the regex of the
+machine types the user wishes to update. The user may also specify by namespace
+or label-selector, or restart the VMs immediately are using the `--namespace`,
+`--label-selector`, and `--restart-now` respectively. In follow-ups to the initial
+implementation, other methods of specifying which VMs to convert the machine
+types of will be added, such as specifying a single VM by name, or configuring
+a limit on the number of VMs that can be restarted at once when restarting the
+running VMs immediately.
+
+Using the CentOS Stream example from the overview, given a VM with the following spec:
+```yaml
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+...
+spec:
+  running: false
+  template:
+    spec:
+      domain:
+        machine:
+          type: pc-q35-rhel8.2.0
+```
+When the user calls the virtctl command:
+`virtctl update machine-types --which-matches-glob *rhel-8.*`
+
+The MMTT Job is deployed, and the VM spec will be updated to: 
+```yaml
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+...
+spec:
+  running: false
+  template:
+    spec:
+      domain:
+        machine:
+          type:
+```
+where the machine type in the VM spec is now empty. Upon starting this VM,
+the machine type will automatically be populated with the default machine type
+in the Kubevirt CR.
+
+A VM may have a machine type of `q35`, an alias for the current latest machine type *at the time the VM is started*.
+
+For example, this VM:
+```yaml
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+...
+spec:
+  running: true
+  template:
+    spec:
+      domain:
+        machine:
+          type: q35
+```
+may have been started when a RHEL-8 machine type was the latest machine type and
+not been stopped or restarted since. Its VMI would thus look like this:
+```yaml
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstance
+...
+spec:
+  domain:
+    machine:
+      type: q35
+status:
+  machine:
+    type: pc-q35-rhel8.2.0
+```
+and need to have its machine type updated.
+
+To handle this, any running VMs with 'q35' machine type and a machine type that matches
+the regex (reported in VMI's status.machine.type) will be labelled with `restart-vm-required`:
+```yaml
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+	labels: "restart-vm-required"="true"
```

Comment:
Everything in KubeVirt is normally designed to not change metadata or spec. One reason for that is to avoid reconcile loops. For example when people are hosting their objects in git and a process is periodically checking if the repo state is in sync with the cluster.

I can understand the desire to "just" selected with an efficient kubectl query the relevant VMs. This seems to be however another case where we would actually finally want field selector support on CRDs, so that we can put info like this in the status.

Thoughts by others?

Reply:
Good point, I have updated the design proposal to create and use a field in the VM Status rather than updating the VM's labels.